File: contrib/RealTimeHumanSeg/cpp/humanseg_postprocess.cc
//   Copyright (c) 2020 PaddlePaddle Authors. All Rights Reserved.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include <iostream>
#include <string>

#include <opencv2/core/core.hpp>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/optflow.hpp>

#include "humanseg_postprocess.h" // NOLINT

int HumanSegTrackFuse(const cv::Mat &track_fg_cfd,
                      const cv::Mat &dl_fg_cfd,
                      const cv::Mat &dl_weights,
                      const cv::Mat &is_track,
                      const float cfd_diff_thres,
                      const int patch_size,
                      cv::Mat cur_fg_cfd) {
  float *cur_fg_cfd_ptr = reinterpret_cast<float*>(cur_fg_cfd.data);
  float *dl_fg_cfd_ptr = reinterpret_cast<float*>(dl_fg_cfd.data);
  float *track_fg_cfd_ptr = reinterpret_cast<float*>(track_fg_cfd.data);
  float *dl_weights_ptr = reinterpret_cast<float*>(dl_weights.data);
  uchar *is_track_ptr = reinterpret_cast<uchar*>(is_track.data);
  int y_offset = 0;
  int ptr_offset = 0;
  int h = track_fg_cfd.rows;
  int w = track_fg_cfd.cols;
  float dl_fg_score = 0.0;
  float track_fg_score = 0.0;
  for (int y = 0; y < h; ++y) {
    for (int x = 0; x < w; ++x) {
      dl_fg_score = dl_fg_cfd_ptr[ptr_offset];
      if (is_track_ptr[ptr_offset] > 0) {
        track_fg_score = track_fg_cfd_ptr[ptr_offset];
        if (dl_fg_score > 0.9 || dl_fg_score < 0.1) {
          if (dl_weights_ptr[ptr_offset] <= 0.10) {
            cur_fg_cfd_ptr[ptr_offset] = dl_fg_score * 0.3
                + track_fg_score * 0.7;
          } else {
            cur_fg_cfd_ptr[ptr_offset] = dl_fg_score * 0.4
                + track_fg_score * 0.6;
          }
        } else {
          cur_fg_cfd_ptr[ptr_offset] = dl_fg_score * dl_weights_ptr[ptr_offset]
            + track_fg_score * (1 - dl_weights_ptr[ptr_offset]);
        }
      } else {
        cur_fg_cfd_ptr[ptr_offset] = dl_fg_score;
      }
      ++ptr_offset;
    }
    y_offset += w;
    ptr_offset = y_offset;
  }
  return 0;
}

int HumanSegTracking(const cv::Mat &prev_gray,
                     const cv::Mat &cur_gray,
                     const cv::Mat &prev_fg_cfd,
                     int patch_size,
                     cv::Mat track_fg_cfd,
                     cv::Mat is_track,
                     cv::Mat dl_weights,
                     cv::Ptr<cv::optflow::DISOpticalFlow> disflow) {
  cv::Mat flow_fw;
  disflow->calc(prev_gray, cur_gray, flow_fw);

  cv::Mat flow_bw;
  disflow->calc(cur_gray, prev_gray, flow_bw);

  float double_check_thres = 8;

  cv::Point2f fxy_fw;
  int dy_fw = 0;
  int dx_fw = 0;
  cv::Point2f fxy_bw;
  int dy_bw = 0;
  int dx_bw = 0;

  float *prev_fg_cfd_ptr = reinterpret_cast<float*>(prev_fg_cfd.data);
  float *track_fg_cfd_ptr = reinterpret_cast<float*>(track_fg_cfd.data);
  float *dl_weights_ptr = reinterpret_cast<float*>(dl_weights.data);
  uchar *is_track_ptr = reinterpret_cast<uchar*>(is_track.data);

  int prev_y_offset = 0;
  int prev_ptr_offset = 0;
  int cur_ptr_offset = 0;
  float *flow_fw_ptr = reinterpret_cast<float*>(flow_fw.data);

  float roundy_fw = 0.0;
  float roundx_fw = 0.0;
  float roundy_bw = 0.0;
  float roundx_bw = 0.0;

  int h = prev_fg_cfd.rows;
  int w = prev_fg_cfd.cols;
  for (int r = 0; r < h; ++r) {
    for (int c = 0; c < w; ++c) {
      ++prev_ptr_offset;

      fxy_fw = flow_fw.ptr<cv::Point2f>(r)[c];
      roundy_fw = fxy_fw.y >= 0 ? 0.5 : -0.5;
      roundx_fw = fxy_fw.x >= 0 ? 0.5 : -0.5;
      dy_fw = static_cast<int>(fxy_fw.y + roundy_fw);
      dx_fw = static_cast<int>(fxy_fw.x + roundx_fw);

      int cur_x = c + dx_fw;
      int cur_y = r + dy_fw;

      if (cur_x < 0
          || cur_x >= h
          || cur_y < 0
          || cur_y >= w) {
        continue;
      }
      fxy_bw = flow_bw.ptr<cv::Point2f>(cur_y)[cur_x];
      roundy_bw = fxy_bw.y >= 0 ? 0.5 : -0.5;
      roundx_bw = fxy_bw.x >= 0 ? 0.5 : -0.5;
      dy_bw = static_cast<int>(fxy_bw.y + roundy_bw);
      dx_bw = static_cast<int>(fxy_bw.x + roundx_bw);

      auto total = (dy_fw + dy_bw) * (dy_fw + dy_bw)
                        + (dx_fw + dx_bw) * (dx_fw + dx_bw);
      if (total >= double_check_thres) {
        continue;
      }

      cur_ptr_offset = cur_y * w + cur_x;
      if (abs(dy_fw) <= 0
          && abs(dx_fw) <= 0
          && abs(dy_bw) <= 0
          && abs(dx_bw) <= 0) {
        dl_weights_ptr[cur_ptr_offset] = 0.05;
      }
      is_track_ptr[cur_ptr_offset] = 1;
      track_fg_cfd_ptr[cur_ptr_offset] = prev_fg_cfd_ptr[prev_ptr_offset];
    }
    prev_y_offset += w;
    prev_ptr_offset = prev_y_offset - 1;
  }
  return 0;
}

int MergeProcess(const uchar *im_buff,
                const float *scoremap_buff,
                const int height,
                const int width,
                uchar *result_buff) {
  cv::Mat prev_fg_cfd;
  cv::Mat cur_fg_cfd;
  cv::Mat cur_fg_mask;
  cv::Mat track_fg_cfd;
  cv::Mat prev_gray;
  cv::Mat cur_gray;
  cv::Mat bgr_temp;
  cv::Mat is_track;
  cv::Mat static_roi;
  cv::Mat weights;
  cv::Ptr<cv::optflow::DISOpticalFlow> disflow =
      cv::optflow::createOptFlow_DIS(
          cv::optflow::DISOpticalFlow::PRESET_ULTRAFAST);

  bool is_init = false;
  const float *cfd_ptr = scoremap_buff;
  if (!is_init) {
    is_init = true;
    cur_fg_cfd = cv::Mat(height, width, CV_32FC1, cv::Scalar::all(0));
    memcpy(cur_fg_cfd.data, cfd_ptr, height * width * sizeof(float));
    cur_fg_mask = cv::Mat(height, width, CV_8UC1, cv::Scalar::all(0));

    if (height <= 64 || width <= 64) {
        disflow->setFinestScale(1);
    } else if (height <= 160 || width <= 160) {
        disflow->setFinestScale(2);
    } else {
        disflow->setFinestScale(3);
    }
    is_track = cv::Mat(height, width, CV_8UC1, cv::Scalar::all(0));
    static_roi = cv::Mat(height, width, CV_8UC1, cv::Scalar::all(0));
    track_fg_cfd = cv::Mat(height, width, CV_32FC1, cv::Scalar::all(0));

    bgr_temp = cv::Mat(height, width, CV_8UC3);
    memcpy(bgr_temp.data, im_buff, height * width * 3 * sizeof(uchar));
    cv::cvtColor(bgr_temp, cur_gray, cv::COLOR_BGR2GRAY);
    weights = cv::Mat(height, width, CV_32FC1, cv::Scalar::all(0.30));
  } else {
    memcpy(cur_fg_cfd.data, cfd_ptr, height * width * sizeof(float));
    memcpy(bgr_temp.data, im_buff, height * width * 3 * sizeof(uchar));
    cv::cvtColor(bgr_temp, cur_gray, cv::COLOR_BGR2GRAY);
    memset(is_track.data, 0, height * width * sizeof(uchar));
    memset(static_roi.data, 0, height * width * sizeof(uchar));
    weights = cv::Mat(height, width, CV_32FC1, cv::Scalar::all(0.30));
    HumanSegTracking(prev_gray,
                     cur_gray,
                     prev_fg_cfd,
                     0,
                     track_fg_cfd,
                     is_track,
                     weights,
                     disflow);
    HumanSegTrackFuse(track_fg_cfd,
                      cur_fg_cfd,
                      weights,
                      is_track,
                      1.1,
                      0,
                      cur_fg_cfd);
  }
  int ksize = 3;
  cv::GaussianBlur(cur_fg_cfd, cur_fg_cfd, cv::Size(ksize, ksize), 0, 0);
  prev_fg_cfd = cur_fg_cfd.clone();
  prev_gray = cur_gray.clone();
  cur_fg_cfd.convertTo(cur_fg_mask, CV_8UC1, 255);
  memcpy(result_buff, cur_fg_mask.data, height * width);
  return 0;
}

cv::Mat MergeSegMat(const cv::Mat& seg_mat,
                    const cv::Mat& ori_frame) {
  cv::Mat return_frame;
  cv::resize(ori_frame, return_frame, cv::Size(ori_frame.cols, ori_frame.rows));
  for (int i = 0; i < ori_frame.rows; i++) {
    for (int j = 0; j < ori_frame.cols; j++) {
      float score = seg_mat.at<uchar>(i, j) / 255.0;
      if (score > 0.1) {
        return_frame.at<cv::Vec3b>(i, j)[2] = static_cast<int>((1 - score) * 255
            + score*return_frame.at<cv::Vec3b>(i, j)[2]);
        return_frame.at<cv::Vec3b>(i, j)[1] = static_cast<int>((1 - score) * 255
            + score*return_frame.at<cv::Vec3b>(i, j)[1]);
        return_frame.at<cv::Vec3b>(i, j)[0] = static_cast<int>((1 - score) * 255
            + score*return_frame.at<cv::Vec3b>(i, j)[0]);
      } else {
        return_frame.at<cv::Vec3b>(i, j) = {255, 255, 255};
      }
    }
  }
  return return_frame;
}

int ThresholdMask(const cv::Mat &fg_cfd,
                  const float fg_thres,
                  const float bg_thres,
                  cv::Mat& fg_mask) {
  if (fg_cfd.type() != CV_32FC1) {
    printf("ThresholdMask: type is not CV_32FC1.\n");
    return -1;
  }
  if (!(fg_mask.type() == CV_8UC1
        && fg_mask.rows == fg_cfd.rows
        && fg_mask.cols == fg_cfd.cols)) {
    fg_mask = cv::Mat(fg_cfd.rows, fg_cfd.cols, CV_8UC1, cv::Scalar::all(0));
  }

  for (int r = 0; r < fg_cfd.rows; ++r) {
    for (int c = 0; c < fg_cfd.cols; ++c) {
      float score = fg_cfd.at<float>(r, c);
      if (score < bg_thres) {
        fg_mask.at<uchar>(r, c) = 0;
      } else if (score > fg_thres) {
        fg_mask.at<uchar>(r, c) = 255;
      } else {
        fg_mask.at<uchar>(r, c) = static_cast<int>(
            (score-bg_thres) / (fg_thres - bg_thres) * 255);
      }
    }
  }
  return 0;
}

File: contrib/HumanSeg/utils/humanseg_postprocess.py
# coding: utf8
# Copyright (c) 2019 PaddlePaddle Authors. All Rights Reserve.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#    http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

import numpy as np
import cv2
import os


def get_round(data):
    round = 0.5 if data >= 0 else -0.5
    return (int)(data + round)


def human_seg_tracking(pre_gray, cur_gray, prev_cfd, dl_weights, disflow):
    """计算光流跟踪匹配点和光流图
    输入参数:
        pre_gray: 上一帧灰度图
        cur_gray: 当前帧灰度图
        prev_cfd: 上一帧光流图
        dl_weights: 融合权重图
        disflow: 光流数据结构
    返回值:
        is_track: 光流点跟踪二值图，即是否具有光流点匹配
        track_cfd: 光流跟踪图
    """
    check_thres = 8
    h, w = pre_gray.shape[:2]
    track_cfd = np.zeros_like(prev_cfd)
    is_track = np.zeros_like(pre_gray)
    flow_fw = disflow.calc(pre_gray, cur_gray, None)
    flow_bw = disflow.calc(cur_gray, pre_gray, None)
    for r in range(h):
        for c in range(w):
            fxy_fw = flow_fw[r, c]
            dx_fw = get_round(fxy_fw[0])
            cur_x = dx_fw + c
            dy_fw = get_round(fxy_fw[1])
            cur_y = dy_fw + r
            if cur_x < 0 or cur_x >= w or cur_y < 0 or cur_y >= h:
                continue
            fxy_bw = flow_bw[cur_y, cur_x]
            dx_bw = get_round(fxy_bw[0])
            dy_bw = get_round(fxy_bw[1])
            if ((dy_fw + dy_bw) * (dy_fw + dy_bw) +
                (dx_fw + dx_bw) * (dx_fw + dx_bw)) >= check_thres:
                continue
            if abs(dy_fw) <= 0 and abs(dx_fw) <= 0 and abs(dy_bw) <= 0 and abs(
                    dx_bw) <= 0:
                dl_weights[cur_y, cur_x] = 0.05
            is_track[cur_y, cur_x] = 1
            track_cfd[cur_y, cur_x] = prev_cfd[r, c]
    return track_cfd, is_track, dl_weights


def human_seg_track_fuse(track_cfd, dl_cfd, dl_weights, is_track):
    """光流追踪图和人像分割结构融合
    输入参数:
        track_cfd: 光流追踪图
        dl_cfd: 当前帧分割结果
        dl_weights: 融合权重图
        is_track: 光流点匹配二值图
    返回
        cur_cfd: 光流跟踪图和人像分割结果融合图
    """
    fusion_cfd = dl_cfd.copy()
    idxs = np.where(is_track > 0)
    for i in range(len(idxs[0])):
        x, y = idxs[0][i], idxs[1][i]
        dl_score = dl_cfd[x, y]
        track_score = track_cfd[x, y]
        fusion_cfd[x, y] = dl_weights[x, y] * dl_score + (
            1 - dl_weights[x, y]) * track_score
        if dl_score > 0.9 or dl_score < 0.1:
            if dl_weights[x, y] < 0.1:
                fusion_cfd[x, y] = 0.3 * dl_score + 0.7 * track_score
            else:
                fusion_cfd[x, y] = 0.4 * dl_score + 0.6 * track_score
        else:
            fusion_cfd[x, y] = dl_weights[x, y] * dl_score + (
                1 - dl_weights[x, y]) * track_score
    return fusion_cfd


def postprocess(cur_gray, scoremap, prev_gray, pre_cfd, disflow, is_init):
    """光流优化
    Args:
        cur_gray : 当前帧灰度图
        pre_gray : 前一帧灰度图
        pre_cfd  ：前一帧融合结果
        scoremap : 当前帧分割结果
        difflow  : 光流
        is_init : 是否第一帧
    Returns:
        fusion_cfd : 光流追踪图和预测结果融合图
    """
    height, width = scoremap.shape[0], scoremap.shape[1]
    disflow = cv2.DISOpticalFlow_create(cv2.DISOPTICAL_FLOW_PRESET_ULTRAFAST)
    h, w = scoremap.shape
    cur_cfd = scoremap.copy()

    if is_init:
        is_init = False
        if h <= 64 or w <= 64:
            disflow.setFinestScale(1)
        elif h <= 160 or w <= 160:
            disflow.setFinestScale(2)
        else:
            disflow.setFinestScale(3)
        fusion_cfd = cur_cfd
    else:
        weights = np.ones((w, h), np.float32) * 0.3
        track_cfd, is_track, weights = human_seg_tracking(
            prev_gray, cur_gray, pre_cfd, weights, disflow)
        fusion_cfd = human_seg_track_fuse(track_cfd, cur_cfd, weights, is_track)

    fusion_cfd = cv2.GaussianBlur(fusion_cfd, (3, 3), 0)

    return fusion_cfd


def threshold_mask(img, thresh_bg, thresh_fg):
    dst = (img / 255.0 - thresh_bg) / (thresh_fg - thresh_bg)
    dst[np.where(dst > 1)] = 1
    dst[np.where(dst < 0)] = 0
    return dst.astype(np.float32)
